Seminar 8 DNA methylation analysis with Illumina's Infinium HumanMethylation450K array
========================================================

Ref1: http://www.ugrad.stat.ubc.ca/~stat540/seminars/seminar08_methylation.html


*Excerpt from ref*
This seminar is inspired by a STAT 540 project in 2013: Analysis of Gene Expression Omnibus Leukaemia Data from the Illumina HumanMethylation450 Array by Alice Zhu, Rachel Edgar, Shaun Jackman and Nick Fishbane. See their project website here.

```{r}
  library(GEOquery)
  library(gplots)
  library(wateRmelon)
  library(IlluminaHumanMethylation450k.db)
  library(reshape2)
  library(lattice)
  require(gridExtra) # also loads grid
  library(limma)
```

```{r}
  source("style.R")
```

## Explore 450k methylation array data

Download/load the dataset
```{r}
  setwd("/Users/sohrab/Me/Apply/Canada Apply/Courses/Second Semester/Stat 540/Lab/stat540-2014-salehi-sohrab/data")

  if (file.exists("methyl_ALL.Rdata")) {
    # if previously downloaded
    load("methyl_ALL.Rdata")
} else {
    # if downloading for the first time
    GSE39141 <- getGEO("GSE39141")
    show(GSE39141)  ## 33 samples (29 ALL and 4 healthy B cells)
    GSE42865 <- getGEO("GSE42865")  # took ~2 mins for JB
    show(GSE42865)  ## 16 samples (9 healthy cells B cells and 7 other cells)

    # Extract expression matrices (turn into data frames at once)
    ALL.dat <- as.data.frame(exprs(GSE39141[[1]]))
    CTRL.dat <- as.data.frame(exprs(GSE42865[[1]]))

    # Obtain the meta-data for the samples and rename them perhaps?
    ALL.meta <- pData(phenoData(GSE39141[[1]]))
    CTRL.meta <- pData(phenoData(GSE42865[[1]]))

    # create some labels
    ALL.meta$Group <- c(rep("ALL", 29), rep("HBC", 4))
    ## ALL: Case; HBC: Healthy B Cells

    # Subset both meta-data and data for control (healthy) donors
    CTRL.meta <- droplevels(subset(CTRL.meta, grepl("Healthy donor", characteristics_ch1.1)))
    CTRL.dat <- subset(CTRL.dat, select = as.character(CTRL.meta$geo_accession))

    # Rename variables
    names(ALL.dat) <- paste(ALL.meta$Group, gsub("GSM", "", names(ALL.dat)), 
        sep = "_")
    names(CTRL.dat) <- paste("HBC", gsub("GSM", "", names(CTRL.dat)), sep = "_")

    # save the data to avoid future re-downloading
    save(ALL.dat, CTRL.dat, ALL.meta, CTRL.meta, file = "methyl_ALL.Rdata")
}
```
Density plot of average Beta values for probes in the two datasets:
```{r}
  # package all together
  t <- melt(CTRL.dat)
  t1 <- melt(ALL.dat)
  t2 <- rbind(t, t1)
  t2$condition  <- factor(rep(c("CTRL", "ALL"), times = c(nrow(t), nrow(t1))))

  densityplot(~value, t2, group = condition, auto.key = T, plot.points = F, xlab="Beta") 


  (p <- ggplot(fDat, aes(geneExp, BrainRegion, color=Genotype), main="Hello") +
  geom_point() +
  facet_wrap(~Sex) + 
  labs(title = paste("Gene expression for probeset", probeName)) +
  xlab("Gene Expression") +
  ylab("Brain Region"))

```


## Normalization

Using `wateRmelon`, quantile normalize the experiments via `betaqn` method. There're 15 different normalization methods. Let's see what [this](http://www.biomedcentral.com/1471-2164/14/293) paper has to say...

$Î² = M/(M + U + 100)$ where $M$ indicates intensity of methylated and $U$ not methylated. They conclude that quantile normalization yeilds `markedly` improved methods, using the 3 metrics they've derived.

Check out [this paper](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3012676/) for more info on $\beta$ and $M$ values.

```{r}
  beta.matrix <- as.matrix(cbind(ALL.dat, CTRL.dat))
  str(beta.matrix, max.level = 0)
  system.time(beta.norm <- betaqn(beta.matrix))

  t3 <- melt(beta.norm)
  t3$condition <- t2$condition 
  t3$Var2 <- NULL

  t2.sample <- t2[sample(seq(nrow(t2)), 100000), ] 
  t3.sample <- t3[sample(seq(nrow(t3)), 100000), ]
  

  p1 <- densityplot(~value, t2.sample, group = condition, auto.key = T, plot.points = F, xlab="Beta", main="Before") 
  p2 <- densityplot(~value, t3.sample, group = condition, auto.key = T, plot.points = F, xlab="Beta", main="After") 
  
  grid.arrange(p1,p2, nrow = 2)
```

The above plots look more similar to me.


## M values

Assumptions of a linear model: don't have more than one mode and span $R$?
<br />
[Here's](http://people.duke.edu/~rnau/testing.htm) a reference on assumption of linear regression. The most relevant I've found is the normality of the errors. So is $\Beta$ value indicative of some kind of an error?


```{r}
  # we'll use this normalization for the probes from now on
  M.norm <- beta2m(beta.norm)
```

## CpG Islands
The idea is that CpG islands have a better understood biological function. It'd be interesting to know whether the mean
of the probes in the same CpG island is a good enough representation of it's methylation state or not...
```{r}
  # Annotate probe ID to CpG islands association
  cginame <- as.data.frame(IlluminaHumanMethylation450kCPGINAME)
  names(cginame) <- c("Probe_ID", "cginame")
  rownames(cginame) <- cginame$Probe_ID
  length(levels(factor(cginame$cginame)))  # No. of CGIs

  head(cginame)
```
Now we'll remove those we don't have a CpG island for:
```{r}
  # restrict probes to those within CGIs
  beta.inCGI <- beta.norm[cginame$Probe_ID, ] # restricting and rearranging
  M.inCGI <- M.norm[cginame$Probe_ID, ]
  nrow(M.inCGI)  # No. of probes within CGIs
```
Now, for each CpG island, assign the mean of probe beta.norm:
```{r}
  # aggregate probes to CGIs

  # how is this working? The elements of cginame$chiname do not seem to match those of the 
  # beta.inCGI. 
  beta.CGI <- aggregate(beta.inCGI, by = list(cginame$cginame), mean, na.rm = T)

  # little peek on how it works. I've a larger test at the end of the file. The ?aggregate wasn't helpful. and couldn't
  # find the binary 
  length(cginame$cginame) == nrow(beta.inCGI)
  any(cginame$cginame %in% rownames(beta.inCGI))
  nrow(beta.inCGI)
  nrow(beta.CGI)
  
  rownames(beta.CGI) <- beta.CGI[, "Group.1"]
  beta.CGI <- subset(beta.CGI, select = -Group.1)
  str(beta.CGI, max.level = 0)

  M.CGI <- aggregate(M.inCGI, by = list(cginame$cginame), mean, na.rm = T)
  rownames(M.CGI) <- M.CGI[, "Group.1"]
  M.CGI <- subset(M.CGI, select = -Group.1)
  str(M.CGI, max.level = 0)
```

Now make a boxplot of the m-values across samples:
```{r}
  molten <- melt(M.CGI)
  molten$group <- factor(     rep(c("ALL", "HBC"), time = c(29 * nrow(M.CGI), 13 * nrow(M.CGI))) , levels=( c("ALL", "HBC")) )
  head(molten)
  p <- ggplot((molten), aes(x=variable, y=value)) + 
                geom_boxplot(aes(fill=factor(group))) + 
                theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
                labs(title = "Distribution of CGI M values", x="", y="M-Values")
  print(p)
   #scale_fill_manual(values = c("white","red")) 
   #theme(axis.text.x = element_text(angle = 90, hjust = 1))
```


## Differential methylation analysis with limma
Are we taking the size of a CpG island into account with this method?
```{r}
  design <- data.frame(Group = relevel(factor(gsub("_[0-9]+", "", colnames(M.CGI))), 
  ref = "HBC"), row.names = colnames(M.CGI))
  str(design)
  (DesMat <- model.matrix(~Group, design))
  DMRfit <- lmFit(M.CGI, DesMat)
  DMRfitEb <- eBayes(DMRfit)
  cutoff <- 0.01
  DMR <- topTable(DMRfitEb, coef = "GroupALL", number = Inf, p.value = cutoff)
  head(DMR)  # top hits 
```

Now visualization, starting with a nice heatmap:
```{r}
  DMR100 <- topTable(DMRfitEb, coef = "GroupALL", number = 100)
  DMR.CGI <- t(as.matrix(subset(beta.CGI, rownames(beta.CGI) %in% rownames(DMR100))))
  str(DMR.CGI, max.level = 0)
  col <- c(rep("darkgoldenrod1", times = nrow(DMR.CGI)))
  col[grepl("HBC", rownames(DMR.CGI))] <- "forestgreen"
  op <- par(mai = rep(0.5, 4))
  heatmap.2(DMR.CGI, col = redblue(256), RowSideColors = col, density.info = "none", 
      trace = "none", Rowv = TRUE, Colv = TRUE, labCol = FALSE, labRow = FALSE, 
      dendrogram = "row", margins = c(1, 5))
  legend("topright", c("ALL", "HBC"), col = c("darkgoldenrod1", "forestgreen"), 
    pch = 15)
```

```{r}
  par(op)
  DMR5 <- topTable(DMRfitEb, coef = "GroupALL", number = 5)
  beta.DMR5probe <- beta.inCGI[cginame[rownames(beta.inCGI), ]$cginame %in% rownames(DMR5), 
    ]
  beta.DMR5probe.tall <- melt(beta.DMR5probe, value.name = "M", varnames = c("Probe_ID", 
    "Sample"))
  beta.DMR5probe.tall$Group <- factor(gsub("_[0-9]+", "", beta.DMR5probe.tall$Sample))
  beta.DMR5probe.tall$CGI <- factor(cginame[as.character(beta.DMR5probe.tall$Probe_ID), 
    ]$cginame)
  (beta.DMR5.stripplot <- ggplot(data = beta.DMR5probe.tall, aes(x = Group, y = M, 
    color = Group)) + geom_point(position = position_jitter(width = 0.05), na.rm = T) + 
    stat_summary(fun.y = mean, aes(group = 1), geom = "line", color = "black") + 
    facet_grid(. ~ CGI) + ggtitle("Probe beta values within top 5 DM CGIs") + 
    xlab("Group") + ylab("beta") + theme_bw())
```

Location of probes along different chromosomes. This is a factastic plot! Could be used in different plots, like
visualizing 4C data:
```{r}
  # get the length of chromosome 1-22 and X
  chrlen <- unlist(as.list(IlluminaHumanMethylation450kCHRLENGTHS)[c(as.character(1:22), "X")])
  chrlen <- data.frame(chr = factor(names(chrlen)), length = chrlen)
  chr <- IlluminaHumanMethylation450kCHR  # get the chromosome of each probe

  # get the probe identifiers that are mapped to chromosome
  chr <- unlist(as.list(chr[mappedkeys(chr)]))

  # get chromosome coordinate of each probe
  coord <- IlluminaHumanMethylation450kCPGCOORDINATE

  # get the probe identifiers that are mapped to coordinate
  coord <- unlist(as.list(coord[mappedkeys(coord)]))
  coord <- data.frame(chr = chr[intersect(names(chr), names(coord))], coord = coord[intersect(names(chr), names(coord))])

  # coordinates of probes in DM CGIs
  coordDMRprobe <- droplevels(na.omit(coord[cginame[cginame$cginame %in% rownames(DMR), ]$Probe_ID, ]))
  (coord.plot <- ggplot(data = coordDMRprobe) + geom_linerange(aes(factor(chr, 
      levels = c("X", as.character(22:1))), ymin = 0, ymax = length), data = chrlen, 
      alpha = 0.5) + geom_point(aes(x = factor(chr, levels = c("X", as.character(22:1))), 
      y = coord), position = position_jitter(width = 0.03), na.rm = T) + ggtitle("DMR positions on chromosomes") + 
      ylab("Position of DMRs") + xlab("chr") + coord_flip() + theme_bw())
```



Little test to see how exactly aggregate works:
```{r}
  
  set.seed(1)
  l1 <- lapply(seq(20), function(x) { paste("A", x, sep="")  }  )
  l2 <- lapply(seq(4), function(x) { paste("S", x, sep="")  }  )
  l3 <- (sample(l2, 20, replace=T))
  l3 <- unlist(l3)
  
  the.mapping <- data.frame(prob=unlist(l1), set=l3)
  
  tat <- unlist(sample(l1, 20, replace = F))
  values <- runif(20)
  

  the.val <- data.frame (tat, values)
  the.val
  

  kaka <- aggregate(the.val, by=list(l3), mean, na.rm = T)


  
  f <- table(the.mapping)
  ff <- lapply(levels(factor(l3)), function(x) which(f[, x] > 0))
  names(ff) <- levels(factor(l3))
  ff
  
  nvalues <- 1:20
  # S1
  nvalues[as.vector(ff$S1)] <- 1
  nvalues[as.vector(ff$S2)] <- 2
  nvalues[as.vector(ff$S3)] <- 3
  nvalues[as.vector(ff$S4)] <- 4

  
  new.val <- data.frame(sort(tat), nvalues)
  new.val
  new.val$S <- 1:20
  
  new.val$S[as.vector(ff$S1)] <- "S1"
  new.val$S[as.vector(ff$S2)] <- "S2"
  new.val$S[as.vector(ff$S3)] <- "S3"
  new.val$S[as.vector(ff$S4)] <- "S4"

  new.val
  mean(new.val$nvalues[new.val$S  == "S1"])
  aggregate(new.val[, -c(3)], by=list(l3), mean, na.rm = T)

  # check
  new.val[new.val$tat %in% c("A10", "A11", "A12", "A5"),]



  tt <- data.frame(r1=seq(1:20), r2=c( rep(1, 5), rep(2, 5), rep(3, 5), rep(4, 5)  ))
  tt

  the.na <- c("S1", "S2", "S3", "S4")
  aa <- aggregate(the.val, by=list(the.na), mean, na.rm = T)



  which(f[, "S1"] > 0)


  match(unlist(l1), the.val$tat)

  n <- the.val[match(unlist(l1), the.val$tat), ]

  mean(n[which(f[, "S1"] > 0), ]$value)
  
```

Conclusion:
Not magic, or environment picking. The by argumetn has to be of the same length of the nrow of X.

