Seminar 7 
========================================================

Ref1: http://www.ugrad.stat.ubc.ca/~stat540/#seminar07

*Excerpt from ref*
The mRNA from 21 mice belonging to two different strains have been sequenced

```{r}
library(edgeR)
library(plyr)
library(gplots)
```{r}
  source("style.R")
```

# load the data
```{r}
  setwd("/Users/sohrab/Me/Apply/Canada Apply/Courses/Second Semester/Stat 540/Lab/stat540-2014-salehi-sohrab/Seminar 7")
  dat <- read.table("data/bottomly_count_table.tsv", header = TRUE, row.names = 1)
  des <- read.table("data/bottomly_phenodata.tsv", header = TRUE, row.names = 1)
  str(dat)

# sanity check
  des
  all(rownames(des) == colnames(dat))
```

Take a look at [this](http://will-landau.com/research/jsm2013.pdf) to have a comparison of dispersion computation methods.

Furthermore, take a look at [this](http://en.wikipedia.org/wiki/Trend_estimation#Fitting_a_trend:_least-squares) for finding out what trend means. 

"trend" will mean the slope of the least squares line?

# GLM edgeR
```{r}
# pointer1: http://www.stat.berkeley.edu/~stark/SticiGui/Text/randomVariables.htm
# a video on Negative Binomial

# ref3: http://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf
# Explaining the GLM edgeR,
```
$$ Y \sim NB(\mu, \varphi) \text{ with mean } \mu \text{ and dispersion } \varphi$$

$$ f(y; \mu, \phi) = P(Y=y) = \frac{\Gamma(y+\phi^{-1})}{\Gamma(\phi^{-1})\Gamma(y+1)}(\frac{1}{1+\mu \phi})^{\phi^{-1}}(\frac{\mu}{\phi^{-1} + \mu})^y  $$


According to [this,](https://stat.ethz.ch/pipermail/bioconductor/2014-January/056975.html) the edgeR pipline has been updated, now you can just use:

```{r}
 # y <- estimateDisp(y,design)
```

According to [this](http://comments.gmane.org/gmane.science.biology.informatics.conductor/52384) you should be using trended dispersion in RNA-seq data modeling, for there are some 


(qCML) is only applicable to experiments with 1 factor (basically estimateCommonDisp() and
estimateTagwiseDisp() functions).
```{r}
table(des$strain)
group <- factor(c(rep("1", 10), rep("2", 11)))
group

dge.glm <- DGEList(counts = dat, group = group)
str(dge.glm)
names(dge.glm)
dge.glm[["samples"]]

dge.glm$counts[1:10, 1:3]

length(dge.glm$counts[1,])
length(dge.glm$counts[,1])

# make the design matrix
design <- model.matrix(~group)
#design
#?estimateGLMCommonDisp
```

Worth's noting that ***tag*** here means trasncript/gene. 

# Symbols from:

$$ f(y; \mu, \phi) = P(Y=y) = \frac{\Gamma(y+\Phi^{-1})}{\Gamma(\Phi =^{-1})\Gamma(y+1)}(\frac{1}{1+\mu \Phi}^{\Phi^{-1}})(\frac{\mu}{\phi^{-1} + \mu})^y  $$

```{r}
# estimate the deviance, according to the following formula from 
# http://biostatistics.oxfordjournals.org/content/9/2/321.full
# Small-sample estimation of negative binomial dispersion, with applications to SAGE data
# Robinson and Smyth 2007


# Here we're estimating different types of dispersion. According to [here](https://stat.ethz.ch/pipermail/bioconductor/2014-January/056975.html), 
# this has little to do with the final results. 
# We can just use  y <- estimateDisp(y,design) and it will do these steps itself altogether
# a. assuming all genes/tags share a common dispersion

dge.glm.com.disp <- estimateGLMCommonDisp(dge.glm, design, verbose = TRUE)
stop()
# b. assuming 
dge.glm.trend.disp <- estimateGLMTrendedDisp(dge.glm.com.disp, design)

dge.glm.tag.disp <- estimateGLMTagwiseDisp(dge.glm.trend.disp, design)

# plot the tagwise dispersion against log2-CPM (counts per million)
plotBCV(dge.glm.tag.disp)

fit <- glmFit(dge.glm.tag.disp, design)
colnames(coef(fit))

lrt <- glmLRT(fit, coef = 2)

# like top table?
topTags(lrt)

tt.glm <- topTags(lrt, n = Inf)
class(tt.glm)

nrow(tt.glm$table[tt.glm$table$FDR < 0.01, ])

interestingSamples <- rownames(tt.glm$table[tt.glm$table$FDR < 1e-50, ])
cpm(dge.glm.tag.disp)[interestingSamples, ]

summary(de.glm <- decideTestsDGE(lrt, p = 0.05, adjust = "BH"))

# plotting the tagwise log fold changes against log-cpm (count per million)
tags.glm <- rownames(dge.glm.tag.disp)[as.logical(de.glm)]
plotSmear(lrt, de.tags = tags.glm)
abline(h = c(-2, 2), col = "blue")


```
For home
1. count equal to zero across all samples 

```{r}

  # remove those that 
  dat1 <- dat[rowSums(dat) != 0, ]
  dge.glm <- DGEList(counts = dat1, group = group)
  
  # estimate the tagwise dispersion (and all the other necessary ones)
  y <- estimateDisp(dge.glm, design)
  fit <- glmFit(y, design)
  colnames(coef(fit))

  lrt <- glmLRT(fit, coef = 2)
  topTags(lrt)
```

2. count equal to zero in at least one sample in each genotype group
```{r}

  # remove those that 
  dat2 <- lapply(1:nrow(dat), function(x) all(dat[x,] > 0) )
  dat3 <- dat[unlist(dat2),]
  dge.glm <- DGEList(counts = dat3, group = group)
  
  # estimate the tagwise dispersion (and all the other necessary ones)
  y <- estimateDisp(dge.glm, design)

  fit <- glmFit(y, design)
  colnames(coef(fit))

  lrt <- glmLRT(fit, coef = 2)
  topTags(lrt)
```

#DESeq
```{r}
#source("http://www.bioconductor.org/biocLite.R")
#biocLite("DESeq")

library(DESeq)
deSeqDat <- newCountDataSet(dat, group)
head(counts(deSeqDat))

deSeqDat <- estimateSizeFactors(deSeqDat)
sizeFactors(deSeqDat)

deSeqDat <- estimateDispersions(deSeqDat)
plotDispEsts(deSeqDat)

results <- nbinomTest(deSeqDat, levels(group)[1], levels(group)[2])
str(results)

plotMA(results)

```


# Limma and Voom
## [ref 2: ](http://www.statsci.org/smyth/pubs/VoomPreprint.pdf) explains how a not-count specific method
such as limma could be adapted to count data. Main idea is based on modeling mean-varience relationship.

```{r}
  library(limma)
  norm.factor <- calcNormFactors(dat)
  dat.voomed <- voom(dat, design, plot = TRUE, lib.size = colSums(dat) * norm.factor)

  dat.voomed

  fit <- lmFit(dat.voomed, design)
  fit <- eBayes(fit)
  topTable(fit)
```

# Take home
1. Choose a specific threshold for the adjusted p value, 

```{r}
  cutoff <- 1e-05
```

2. find the genes identified as differentially expressed using each of 

  2.a edgeR, 
```{r}
  group <- factor(c(rep("1", 10), rep("2", 11)))
  dge.glm <- DGEList(counts = dat, group = group)
  design <- model.matrix(~group)

  y <- estimateGLMCommonDisp(dge.glm, design, verbose = TRUE)
  y <- estimateGLMTrendedDisp(y, design)
  y <- estimateGLMTagwiseDisp(y, design)
# it didn't work :(
#  y <- estimateDisp(dge.glm, design)

  fit <- glmFit(y, design)
  colnames(coef(fit))

  lrt <- glmLRT(fit, coef = 2)
  y1 <- topTags(lrt, n=Inf)$table
  y1 <- y1[y1$FDR < cutoff,]
```
  
  2.b DESeq and   
```{r}
  
  deSeqDat <- newCountDataSet(dat, group)
  deSeqDat <- estimateSizeFactors(deSeqDat)
  # add common scale for values in colunms (e.g. samples), this command has a side effect on input
  sizeFactors(deSeqDat)
  
  # why there are NA's in the result?
  deSeqDat <- estimateDispersions(deSeqDat)
  y2 <- nbinomTest(deSeqDat, levels(group)[1], levels(group)[2])
  y2 <- y2[y2$padj < cutoff & !is.na(y2$padj),]
```
  
  2.c voom+limma. 
```{r}
  norm.factor <- calcNormFactors(dat)
  dat.voomed <- voom(dat, design, lib.size = colSums(dat) * norm.factor)

  fit <- lmFit(dat.voomed, design)
  fit <- eBayes(fit)
  y3 <- topTable(fit, p.value = cutoff, n = Inf, coef=2)
  y3 <- y3[y3$adj.P.Val < cutoff,]
```
  
3.a Compare the number of genes in these 3 lists, 
```{r}
  (t = list("edgeR"=nrow(y1), "DESeq"=nrow(y2), "voom+limma"=nrow(y3)))
```
3.b and draw a venn digram demonstrating the overlap (if any!).
```{r}
  venn(list("edgeR"=rownames(y1), "DESeq"=y2$id, "voom+limma"=y3$ID))
```

